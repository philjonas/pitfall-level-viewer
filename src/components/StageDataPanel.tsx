import React from 'react';
import {
  getPitType,
  getPathObject,
  getUnderworld,
  getVine,
  getTreePattern,
} from '../helpers/ScreenAnalyser';

export interface StageDataPanelProps {
  stageData?: number;
}

export const StageDataPanel = ({
  stageData,
}: StageDataPanelProps): JSX.Element => {
  return (
    <div>
      <p data-testid="pit-type">Pit Type: {getPitType(stageData)}</p>
      <p data-testid="path-obj">Path Objects: {getPathObject(stageData)}</p>
      <p data-testid="underworld">Underworld: {getUnderworld(stageData)}</p>
      <p data-testid="is-vine">Vine: {`${getVine(stageData)}`}</p>
      <p data-testid="tree-pat">Tree Pattern: {getTreePattern(stageData)}</p>
    </div>
  );
};
