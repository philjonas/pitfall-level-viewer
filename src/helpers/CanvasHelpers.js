import {
  Tree0,
  Tree1,
  Tree2,
  Tree3,
  Wall_S,
  Wall_R,
  Wall_L,
  Gold,
  Silver,
  Money,
  Ring,
  Path1Hole,
  Path3Holes,
  PathCroc,
  PathSand,
  PathTar,
} from '../images';
import {
  UNDERGROUND_SCORPION,
  UNDERGROUND_WALL_R,
  UNDERGROUND_WALL_L,
  TREASURE_GOLD,
  TREASURE_SILVER,
  TREASURE_MONEY,
  TREASURE_RING,
  OBSTACLE_NONE,
  OBSTACLE_2_A_LOGS,
  OBSTACLE_2_C_LOGS,
  OBSTACLE_R_LOG,
  OBSTACLE_3_R_LOGS,
  OBSTACLE_3_S_LOGS,
  OBSTACLE_FIRE,
  OBSTACLE_SNAKE,
  PIT_1_HOLE,
  PIT_3_HOLES,
  PIT_CROC,
  PIT_QUICK_VINE,
  PIT_SHIFT_QUICK,
  PIT_SHIFT_TAR,
  PIT_TAR_VINE,
  PIT_TREASURE,
  OBSTACLE_S_LOG,
} from '../state/constants';

export const drawToCanvas = (img, canvas) =>
  canvas
    .getContext('2d')
    .drawImage(
      img,
      0,
      0,
      img.width,
      img.height,
      0,
      0,
      canvas.width,
      canvas.height
    );
export const treePatterns = [Tree0, Tree1, Tree2, Tree3];

export const underworldPatterns = {
  [UNDERGROUND_SCORPION]: Wall_S,
  [UNDERGROUND_WALL_R]: Wall_R,
  [UNDERGROUND_WALL_L]: Wall_L,
};

export const pathObjectPatterns = {
  [TREASURE_GOLD]: Gold,
  [TREASURE_SILVER]: Silver,
  [TREASURE_MONEY]: Money,
  [TREASURE_RING]: Ring,
  [OBSTACLE_NONE]: null,
  [OBSTACLE_R_LOG]: null, // TODO: add image later
  [OBSTACLE_S_LOG]: null, // TODO: add image later
  [OBSTACLE_2_A_LOGS]: null, // TODO: add image later
  [OBSTACLE_2_C_LOGS]: null, // TODO: add image later
  [OBSTACLE_3_R_LOGS]: null, // TODO: add image later
  [OBSTACLE_3_S_LOGS]: null, // TODO: add image later
  [OBSTACLE_FIRE]: null, // TODO: add image later
  [OBSTACLE_SNAKE]: null, // TODO: add image later
};

export const pitTypePatterns = {
  [PIT_1_HOLE]: Path1Hole,
  [PIT_3_HOLES]: Path3Holes,
  [PIT_CROC]: PathCroc,
  [PIT_QUICK_VINE]: PathSand,
  [PIT_SHIFT_QUICK]: PathSand,
  [PIT_SHIFT_TAR]: PathTar,
  [PIT_TAR_VINE]: PathTar,
  [PIT_TREASURE]: PathTar,
};
