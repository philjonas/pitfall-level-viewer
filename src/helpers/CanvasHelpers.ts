import {Tree0, Tree1, Tree2, Tree3, Wall_S, Wall_R, Wall_L} from '../images';
import {
  UNDERGROUND_SCORPION,
  UNDERGROUND_WALL_R,
  UNDERGROUND_WALL_L,
} from '../state/constants';

// @ts-ignore
export const drawToCanvas = (img: HTMLImageElement, canvas) =>
  canvas
    .getContext('2d')
    .drawImage(
      img,
      0,
      0,
      img.width,
      img.height,
      0,
      0,
      canvas.width,
      canvas.height
    );
export const treePatterns = [Tree0, Tree1, Tree2, Tree3];

export const underworldPatterns: {[id: string]: HTMLImageElement} = {
  [UNDERGROUND_SCORPION]: Wall_S,
  [UNDERGROUND_WALL_R]: Wall_R,
  [UNDERGROUND_WALL_L]: Wall_L,
};
