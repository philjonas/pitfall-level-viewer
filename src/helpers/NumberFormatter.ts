const pad = (value: string, length: number): string => {
  let str = value;
  while (str.length < length) {
    str = '0' + str;
  }
  return str;
};

export const formatHex = (value: number): string => {
  return pad(value.toString(16).toUpperCase(), 2);
};

export const formatBin = (value: number): string => {
  return pad(value.toString(2), 8);
};

export const formatDec = (value: number): string => {
  return pad(value.toString(), 3);
};
