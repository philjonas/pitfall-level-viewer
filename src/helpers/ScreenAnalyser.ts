import {
  MASK_1_BIT,
  MASK_2_BIT,
  MASK_3_BIT,
  PIT_1_HOLE,
  PIT_3_HOLES,
  PIT_TAR_VINE,
  PIT_QUICK_VINE,
  PIT_TREASURE,
  PIT_CROC,
  UNDERGROUND_SCORPION,
  UNDERGROUND_WALL_R,
  UNDERGROUND_WALL_L,
  OBSTACLE_NONE,
  PIT_TYPES,
  TREASURES,
  PATH_OBJECTS,
} from '../state/constants';

export const getPitType = (value: number): string => {
  const wallMask = (value >> 3) & MASK_3_BIT;
  return PIT_TYPES[wallMask];
};

export const isTreasureScene = (value: number): boolean => {
  return getPitType(value) === PIT_TREASURE;
};

export const isPathObject = (value: number): boolean => {
  const pit = getPitType(value);
  return pit !== PIT_CROC;
};

export const isWall = (value: number): boolean => {
  const pit = getPitType(value);
  return pit === PIT_1_HOLE || pit === PIT_3_HOLES;
};

export const getUnderworld = (value: number): string => {
  if (!isWall(value)) {
    return UNDERGROUND_SCORPION;
  }
  return (value >> 7) & MASK_1_BIT ? UNDERGROUND_WALL_R : UNDERGROUND_WALL_L;
};

export const getPathObject = (value: number): string => {
  const wallMask = value & MASK_3_BIT;

  if (!isPathObject(value)) {
    return OBSTACLE_NONE;
  }

  if (isTreasureScene(value)) {
    return TREASURES[wallMask % TREASURES.length];
  }

  return PATH_OBJECTS[wallMask];
};

export const getTreePattern = (value: number): number => {
  return (value >> 6) & MASK_2_BIT;
};

export const getVine = (value: number): boolean => {
  const pit = getPitType(value);
  if (pit === PIT_TAR_VINE || pit === PIT_QUICK_VINE) {
    return true;
  }
  if (pit === PIT_CROC) {
    return !!((value >> 1) & MASK_1_BIT);
  }

  return false;
};
