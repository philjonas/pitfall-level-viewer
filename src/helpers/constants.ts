export const MASK_1_BIT = 0b1;
export const MASK_2_BIT = 0b11;
export const MASK_3_BIT = 0b111;
export const MASK_1_BYTE = 0xff;

export const TREASURE_MONEY = 'Money bag';
export const TREASURE_SILVER = 'Silver brick';
export const TREASURE_GOLD = 'Gold brick';
export const TREASURE_RING = 'Diamond ring';

export const PIT_1_HOLE = 'Hole with ladder';
export const PIT_3_HOLES =
  'Hole with ladder surrounded by 2 holes without ladders';
export const PIT_TAR_VINE = 'Tar pit with vine';
export const PIT_QUICK_VINE = 'Quicksand with vine';
export const PIT_CROC = 'Crocodile pit'; // plus a vine if bit-1 is 1
export const PIT_TREASURE = 'Shifting tar pit with treasure';
export const PIT_SHIFT_TAR = 'Shifting tar pit with vine';
export const PIT_SHIFT_QUICK = 'Shifting quicksand';

export const UNDERGROUND_SCORPION = 'Scorpion';
export const UNDERGROUND_WALL_R = 'Wall on the right';
export const UNDERGROUND_WALL_L = 'Wall on the left';

export const OBSTACLE_R_LOG = '1 rolling log';
export const OBSTACLE_2_C_LOGS = '2 rolling logs close together';
export const OBSTACLE_2_A_LOGS = '2 rolling logs spaced apart';
export const OBSTACLE_3_R_LOGS = '3 rolling logs';
export const OBSTACLE_S_LOG = '1 stationary log';
export const OBSTACLE_3_S_LOGS = '3 stationary logs';
export const OBSTACLE_FIRE = 'Fire';
export const OBSTACLE_SNAKE = 'Rattlesnake';
export const OBSTACLE_NONE = 'No obstacles';

export const SCREEN_DATA_001 = 0xc4;
export const SCREEN_DATA_255 = 0xe2;

export const PIT_TYPES = [
  PIT_1_HOLE,
  PIT_3_HOLES,
  PIT_TAR_VINE,
  PIT_QUICK_VINE,
  PIT_CROC,
  PIT_TREASURE,
  PIT_SHIFT_TAR,
  PIT_SHIFT_QUICK,
];

export const PATH_OBJECTS = [
  OBSTACLE_R_LOG,
  OBSTACLE_2_C_LOGS,
  OBSTACLE_2_A_LOGS,
  OBSTACLE_3_R_LOGS,
  OBSTACLE_S_LOG,
  OBSTACLE_3_S_LOGS,
  OBSTACLE_FIRE,
  OBSTACLE_SNAKE,
];

export const TREASURES = [
  TREASURE_MONEY,
  TREASURE_SILVER,
  TREASURE_GOLD,
  TREASURE_RING,
];
