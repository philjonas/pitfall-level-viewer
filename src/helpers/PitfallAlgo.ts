import {MASK_1_BYTE, SCREEN_DATA_001} from './constants';

export const left_random = (value: number): number => {
  let r = value & MASK_1_BYTE;
  const t = (r >> 6) ^ (r >> 5) ^ (r >> 4) ^ r;
  r >>= 1;
  r |= t << 7;
  return r & MASK_1_BYTE;
};

export const right_random = (value: number): number => {
  let r = value & MASK_1_BYTE;
  const t = (r >> 7) ^ (r >> 5) ^ (r >> 4) ^ (r >> 3);
  r <<= 1;
  r |= t & 1;
  return r & MASK_1_BYTE;
};

export const wrapAroundScreenIndex = (value: number): number => {
  if (value < 1) {
    return 255;
  } else if (value > 255) {
    return 1;
  }
  return value;
};

export const getScreenDataByIndex = (value: number): number => {
  const formattedIndex = wrapAroundScreenIndex(value);
  let index = 1;
  let data = SCREEN_DATA_001;
  while (index < formattedIndex) {
    data = right_random(data);
    index += 1;
  }
  return data;
};
