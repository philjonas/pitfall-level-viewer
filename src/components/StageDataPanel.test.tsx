import {render} from '@testing-library/react';
import {StageDataPanelProps, StageDataPanel} from './StageDataPanel';
import {SCREEN_DATA_001} from '../helpers/constants';

const setup = (props?: StageDataPanelProps) => {
  const utils = render(<StageDataPanel {...props} />);
  const pitType = utils.getByTestId('pit-type');
  const pathObj = utils.getByTestId('path-obj');
  const underworld = utils.getByTestId('underworld');
  const isVine = utils.getByTestId('is-vine');
  const treePattern = utils.getByTestId('tree-pat');
  return {
    pitType,
    pathObj,
    underworld,
    isVine,
    treePattern,
    ...utils,
  };
};

describe('StageDataPanel', () => {
  it('renders correctly', () => {
    const wrapper = <StageDataPanel stageData={SCREEN_DATA_001} />;
    expect(wrapper).toMatchSnapshot();
  });

  it('returns the correct text for the screen', () => {
    const {pitType, pathObj, underworld, isVine, treePattern} = setup({
      stageData: SCREEN_DATA_001,
    });
    expect(pitType.innerHTML).toBe('Pit Type: Hole with ladder');
    expect(pathObj.innerHTML).toBe('Path Objects: 1 stationary log');
    expect(underworld.innerHTML).toBe('Underworld: Wall on the right');
    expect(isVine.innerHTML).toBe('Is there a vine: false');
    expect(treePattern.innerHTML).toBe('Tree Pattern: 3');
  });
});
