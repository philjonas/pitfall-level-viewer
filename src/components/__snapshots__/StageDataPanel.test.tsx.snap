// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StageDataPanel renders correctly 1`] = `
<StageDataPanel
  stageData={196}
/>
`;
